Remove deadlock from AuthMap Endpoint GC

meyskens opened the pull request:
This changes that a create and delete event from the EndpointManager will also give a "state of the world" update of all current endpoints it has.
This allows the subscribers to check against other endpoints without calling a function that issues a lock when the event was send inside an RLock.

This is then used in the AuthMap GC code instead of doing a call that caused a deadlock.

Fixes: https://github.com/cilium/cilium/issues/29078

```release-note
Fix potential deadlock that results in stale authentication entries in Cilium
```

bimmlerd commented on pkg/endpointmanager/subscribe.go:
```diff
@@ -10,12 +10,16 @@ type Subscriber interface {
-	EndpointCreated(ep *endpoint.Endpoint)
+	// epList is the list of all endpoints that exist after the creation of
+	// the new endpoint.
+	EndpointCreated(ep *endpoint.Endpoint, epList []*endpoint.Endpoint)
```
On the one hand a bit concerned with giving out pointers to all the EPs under a read lock, as it's then possible to change these in an unsafe way. On the other hand, that's been possible already, so it's not much worse.

bimmlerd commented on pkg/endpointmanager/manager.go:
```diff
@@ -587,11 +587,17 @@ func (mgr *endpointManager) HasGlobalCT() bool {
+}
+
+// getEndpointsLocked returns a slice of all endpoints present in endpoint manager.
+// it should only be called when holding a lock on the endpoint manager.
```
```suggestion
// it should only be called when holding the endpoint manager lock.
```

bimmlerd commented on pkg/endpointmanager/subscribe.go:
```diff
@@ -10,12 +10,16 @@ type Subscriber interface {
 	// EndpointCreated is called at the end of endpoint creation.
 	// Implementations must not attempt write operations on the
```
Could add the caveat of not locking the endpoint manager at all, since this suggests it's OK to do read operations, which it clearly isn't

bimmlerd reviewed (commented):
couple of suggestions; but yep I believe this fixes the deadlock :+1:

christarazi commented on pkg/endpointmanager/subscribe.go:
```diff
@@ -10,12 +10,16 @@ type Subscriber interface {
-	EndpointDeleted(ep *endpoint.Endpoint, conf endpoint.DeleteConfig)
+	// epList is the list of all endpoints that exist after the deletion of
+	// the endpoint.
+	EndpointDeleted(ep *endpoint.Endpoint, conf endpoint.DeleteConfig, epList []*endpoint.Endpoint)
```
It seems like a bit of an abuse of this interface to bolt on `epList` when most subscribers are not going to use it. I wonder if instead of making this change to the interface and having to fetch the list of endpoints from the endpointmanager via `getEndpointsLocked`, we could build up a local cache of endpoints within the auth gc code. Upon initialization of auth gc, it can be initialized with the current list of endpoints from endpointmanager and then auth gc subscribes to `EndpointCreate` to further populate its local cache. Then that local cache can be used instead of reaching back out to the endpointmanager.

meyskens committed:
Remove deadlock from AuthMap Endpoint GC

This change keeps a copy of the endpoints map inside the Autentication
GC code. It will be updated on subscribe events,

This is then used in the AuthMap GC code instead of doing a call that
caused a deadlock.

Signed-off-by: Maartje Eyskens <maartje.eyskens@isovalent.com>

mhofstetter commented on pkg/auth/authmap_gc.go:
```diff
@@ -39,7 +39,10 @@ type authMapGarbageCollector struct {
 	ciliumIdentitiesDeleted    map[identity.NumericIdentity]struct{}
 
-	endpointRepo endpointRepository
+	endpointRepo         endpointRepository
```
`endpointRepo` is no longer used -> delete field and parameter in "constructor"

meyskens replied:
good catch!

mhofstetter commented on pkg/auth/authmap_gc.go:
```diff
@@ -405,30 +408,46 @@ func (r *authMapGarbageCollector) cleanupExpiredEntries(_ context.Context) error
+	delete(r.endpointsCache, ep.GetID16())
+	r.endpointsCacheMutex.Unlock()
+
 	// when an endpoint got removed clean the authmap entries
```
Why do we cleanup endpoints immediately after receiving the deletion event? the rest of the authmap garbage collector  works the way that all events are captured first - and the actual GC execution removes them. 

The reason why I'm asking is because this also includes proper reporting of the error back to the timer-based job of the Hive-framework. Instead, only a warning gets logged here.

IMO we should fix and align this - even though this might be out of scope for this PR because it's not related to the actual issue and has been introduced with the EP garbage collection.

But i'd like to see this in a follow up PR if possible.

meyskens replied:
The reason for this is that this change was done in hardening against possible attacks of stale entries in the auth map. Acting on deletion event itself would close the time gap where this is possible.

The error handling aspect of this might be a good follow up, feel free to file an issue for it happy to look into it

mhofstetter commented on pkg/auth/authmap_gc.go:
```diff
@@ -39,7 +39,10 @@ type authMapGarbageCollector struct {
 
-	endpointRepo endpointRepository
+	endpointRepo         endpointRepository
+	endpointsCache       map[uint16]*endpoint.Endpoint
```
We could improve this by only keeping the relevant information in the cache.

Instead of the whole Endpoint we should only keep its SecurityIdentity as value of this map. WDYT?

This way we preprocess the relevant data in the event handler instead of the GC execution itself.

bimmlerd replied:
> Instead of the whole Endpoint we should only keep its SecurityIdentity as value of this map. WDYT?

I'm not sure whether the SecurityIdentity of an EP can change during it's lifetime. If so, that could break stuff if we only keep the ID instead of the full EP reference, I think?

mhofstetter replied:
Good point. Would be great to know anyway - definitely worth a comment if this is the case. Maybe @christarazi can clarify this?

This is a good additional reason to keep the data kept in the GarbageCollector as little as possible - to get rid of these potential side-effects.

meyskens replied:
this would need some kind of "reference counting" and i am indeed not sure on security IDs changing, they can as we deal with pointers. In the first attempt of this i did something similar and got out of sync issues because of it, though i deleted all that code so couldn't tell exactly what i did.

mhofstetter replied:
> this would need some kind of "reference counting"

_Assuming_ that the ep -> security id is stable: Why should it be necessary to count the references?

In the cleanup, it's only necessary to to get all security ids that are in use -> collecting & merging all values from the map `map[uint16]*identity.Identity` (ep -> SID) (therefore the reason to keep it as map and not keeping it "flatten" as the used security ids)

christarazi replied:
Endpoint security identities can certainly change. The most common way is when the user modifies the labels for the underlying pod. So I think keeping a cache of live references allows fetching the most up to date security identity.

mhofstetter replied:
Thanks for the info! Interesting that there's no "EndpointChanged/Updated" function on the subscriber interface.

mhofstetter reviewed (commented):
Nice. Glad we ended up no longer having a dependency on the endpointrepo from the authmap cache at all. :pray: 

Only some non-blocking suggestions that i like to discuss.

bimmlerd commented on pkg/auth/authmap_gc.go:
```diff
@@ -405,30 +408,46 @@ func (r *authMapGarbageCollector) cleanupExpiredEntries(_ context.Context) error
 
 func (r *authMapGarbageCollector) subscribeToEndpointEvents(endpointManager endpointmanager.EndpointManager) {
 	endpointManager.Subscribe(r)
```
Could move the `Subscribe` call into the below critical section unless I'm missing a reason not to. Subscribing before it allows events to come in prior to acquiring the snapshot, and that feels a bit racy (though I can't see an actual bug).

bimmlerd reviewed (approved):
Also solves the deadlock as far as I can tell :+1:

mhofstetter reviewed (approved):
Good to go from my side! Thanks @meyskens

aditighag commented on pkg/endpointmanager/subscribe.go:
```diff
@@ -10,11 +10,15 @@ type Subscriber interface {
 	// EndpointCreated is called at the end of endpoint creation.
 	// Implementations must not attempt write operations on the
 	// EndpointManager from this callback.
+	// This function is being called inside a RLock, so it must not attempt
```
just an fyi: We typically follow the convention - `EndpointCreatedLocked`.
